src/hyperlight_guest_bin/src/host_bridge.rs

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2025  The Hyperlight Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use alloc::string::ToString;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::cell::OnceCell;
+use core::ffi::*;
+use core::sync::atomic::{AtomicU64, Ordering};
+
+use hyperlight_common::flatbuffer_wrappers::function_types::{ParameterValue, ReturnType};
+
+use crate::host_comm::{call_host_function, get_host_function_details};
+
+static CURRENT_TIME: AtomicU64 = AtomicU64::new(0);
+static CAPS: Caps = Caps::new();
+
+// Host capability flags
+struct Caps {
+    host_read: OnceCell<bool>,
+    host_clock: OnceCell<bool>,
+}
+
+unsafe impl Sync for Caps {}
+
+impl Caps {
+    const fn new() -> Self {
+        Self {
+            host_read: OnceCell::new(),
+            host_clock: OnceCell::new(),
+        }
+    }
+
+    #[inline]
+    fn query(name: &str) -> bool {
+        get_host_function_details()
+            .host_functions
+            .unwrap_or_default()
+            .iter()
+            .any(|f| f.function_name == name)
+    }
+
+    #[inline]
+    fn host_read(&self) -> bool {
+        *self.host_read.get_or_init(|| Self::query("HostRead"))
+    }
+
+    #[inline]
+    fn host_clock(&self) -> bool {
+        *self.host_clock.get_or_init(|| Self::query("CurrentTime"))
+    }
+}
+
+#[unsafe(no_mangle)]
+pub extern "C" fn read(fd: c_int, buf: *mut c_void, count: usize) -> isize {
+    // Only stdin (fd=0) and only if HostRead is defined
+    if fd != 0 || !CAPS.host_read() {
+        return 0;
+    }
+
+    let bytes = call_host_function::<Vec<u8>>(
+        "HostRead",
+        Some(vec![ParameterValue::ULong(count as u64)]),
+        ReturnType::VecBytes,
+    )
+    .unwrap_or_default();
+
+    let n = bytes.len();
```

Comment:
Do we want to assert `n <= count`?
A wrongly implemented `HostRead` could cause a buffer overflow